File: v2/src/sandcastle-runner.ts
/**
 * Per-task docker sandbox runner.
 *
 * Spawns `docker run --rm` against the devclaw-sandbox image for each task.
 * The container's ENTRYPOINT runs the python OpenHands runner, which streams
 * one prefixed JSON line per event (`event: {...}`) plus a single terminating
 * line (`result: {...}`). This module:
 *
 *   - Translates an `OpenHandsRequest` into a docker invocation.
 *   - Bind-mounts the host workspace into /workspace and ~/.claude read-only
 *     into /home/agent/.claude (Pro OAuth posture: claude CLI inside sandbox
 *     can read tokens but not write back).
 *   - Streams stdout line-by-line; routes `event:` lines through `onEvent`
 *     and parses the final `result:` line as the OpenHandsResult.
 *   - Refuses to forward ANTHROPIC_API_KEY into the container (same belt
 *     +suspenders the python runner enforces).
 *
 * Container lifecycle: --rm + the per-task --name make destroy-on-exit
 * automatic. No persistent on-host state.
 *
 * This file replaces openhands-runner's "spawn python directly on host" path
 * for production. Tests still inject a stub runner so they don't need docker.
 */

import { spawn } from "node:child_process";
import { randomUUID } from "node:crypto";
import { existsSync } from "node:fs";
import { homedir } from "node:os";
import { resolve } from "node:path";

/**
 * Inputs the runner needs to launch one OpenHands run. Same kinds the MCP
 * tool surface exposes; the runner.py wrapper picks the right system-prompt
 * template per kind.
 */
export type OpenHandsRequest = {
  kind: "implement_feature" | "fix_bug" | "review_repository";
  workspaceDir: string;
  goal: string;
};

/**
 * Terminal verdict from one run. Mirrors the `result: {...}` line shape that
 * runner.py emits — fields are populated on success, plus `agent_output`
 * for debugging when present.
 */
export type OpenHandsResult =
  | {
      status: "ok";
      workspaceDir: string;
      message: string;
      agent_output?: string;
    }
  | { status: "error"; error: string; trace?: string };

/**
 * Event handed to the caller's `onEvent` for every `event:` line the runner
 * emits. Mirrors the python `_emit_event` payload shape verbatim.
 */
export type RunnerEvent = {
  id: string | null;
  type: string;
  source: string;
  ts: number | string;
  payload: unknown;
};

export type SandcastleRunRequest = OpenHandsRequest & {
  onEvent?: (event: RunnerEvent) => void;
};

const SANDBOX_IMAGE =
  process.env["DEVCLAW_SANDBOX_IMAGE"] ?? "devclaw-sandbox:latest";
const DOCKER_BIN = process.env["DEVCLAW_DOCKER_BIN"] ?? "docker";
// Container-side mount targets. Match Dockerfile expectations.
const CONTAINER_WORKSPACE = "/workspace";
const CONTAINER_CLAUDE_DIR = "/home/agent/.claude";

class SandcastleRunnerError extends Error {
  constructor(message: string, public readonly trace?: string) {
    super(message);
    this.name = "SandcastleRunnerError";
  }
}

/**
 * Run one task inside a fresh sandbox container. Resolves with the same
 * shape as `runOpenHands` so it's a drop-in for TaskQueue.
 */
export async function runSandcastle(
  req: SandcastleRunRequest,
): Promise<OpenHandsResult> {
  const claudeDir =
    process.env["DEVCLAW_HOST_CLAUDE_DIR"] ?? resolve(homedir(), ".claude");
  if (!existsSync(claudeDir)) {
    return {
      status: "error",
      error:
        `host claude dir not found at ${claudeDir} — set DEVCLAW_HOST_CLAUDE_DIR ` +
        `to override or mount your Claude Code config into the devclaw container.`,
    };
  }

  // When devclaw-mcp itself runs in a container and spawns docker on the
  // host socket, the workspace path it sees internally (e.g.
  // /var/lib/devclaw/workspaces/<id>) is not the same as the host's view of
  // that bind-mounted directory (e.g. /srv/devclaw/workspaces/<id>). Docker
  // needs the host path. The path-prefix env pair tells us how to translate.
  // Unset → pass through (typical local dev where we run directly on host).
  const hostBindPath = translateWorkspacePath(req.workspaceDir);

  // Per-task container name for greppable logs + manual cleanup if --rm
  // somehow fails (which it shouldn't, but operators appreciate the hook).
  const containerName = `devclaw-${randomUUID().slice(0, 8)}`;

  const payload = JSON.stringify({
    kind: req.kind,
    workspace_dir: CONTAINER_WORKSPACE,
    goal: req.goal,
  });

  const dockerArgs: string[] = [
    "run",
    "--rm",
    "--name",
    containerName,
    "--network",
    "host", // claude OAuth refresh needs egress; tighten later via allowlist.
    "-v",
    `${hostBindPath}:${CONTAINER_WORKSPACE}`,
    "-v",
    `${claudeDir}:${CONTAINER_CLAUDE_DIR}:ro`,
    "-e",
    "OPENHANDS_SUPPRESS_BANNER=1",
    SANDBOX_IMAGE,
    payload,
  ];

  return new Promise<OpenHandsResult>((resolveResult) => {
    const child = spawn(DOCKER_BIN, dockerArgs, {
      stdio: ["ignore", "pipe", "pipe"],
      env: stripApiKeys(process.env),
    });

    let stdoutBuf = "";
    let stderrBuf = "";
    let result: OpenHandsResult | null = null;

    child.stdout.on("data", (chunk: Buffer) => {
      stdoutBuf += chunk.toString("utf8");
      // Drain complete lines. Anything past the last newline stays buffered.
      const lines = stdoutBuf.split("\n");
      stdoutBuf = lines.pop() ?? "";
      for (const raw of lines) {
        const line = raw.trim();
        if (!line) continue;
        if (line.startsWith("event: ")) {
          if (req.onEvent) {
            try {
              const ev = JSON.parse(line.slice("event: ".length)) as RunnerEvent;
              req.onEvent(ev);
            } catch (parseErr) {
              process.stderr.write(
                `sandcastle-runner: dropping malformed event line: ${(parseErr as Error).message}\n`,
              );
            }
          }
        } else if (line.startsWith("result: ")) {
          // First result line wins; ignore anything after.
          if (result === null) {
            try {
              result = JSON.parse(line.slice("result: ".length)) as OpenHandsResult;
            } catch (parseErr) {
              result = {
                status: "error",
                error: `runner emitted unparsable result: ${(parseErr as Error).message}`,
                trace: line,
              };
            }
          }
        }
        // Everything else is sandbox decorative output — drop.
      }
    });

    child.stderr.on("data", (chunk: Buffer) => {
      stderrBuf += chunk.toString("utf8");
    });

    child.on("error", (err) => {
      resolveResult({
        status: "error",
        error:
          `failed to spawn ${DOCKER_BIN}: ${err.message}. ` +
          `Is docker installed and the socket reachable from this process?`,
      });
    });

    child.on("close", (code) => {
      if (result) {
        resolveResult(result);
        return;
      }
      resolveResult({
        status: "error",
        error:
          `sandbox exited ${code} without a result line. ` +
          `stderr tail:\n${stderrBuf.slice(-1024)}`,
      });
    });
  });
}

function translateWorkspacePath(workspaceDir: string): string {
  const containerPrefix = process.env["DEVCLAW_CONTAINER_PATH_PREFIX"];
  const hostPrefix = process.env["DEVCLAW_HOST_PATH_PREFIX"];
  if (
    containerPrefix &&
    hostPrefix &&
    workspaceDir.startsWith(containerPrefix)
  ) {
    return hostPrefix + workspaceDir.slice(containerPrefix.length);
  }
  return workspaceDir;
}

function stripApiKeys(
  env: NodeJS.ProcessEnv,
): NodeJS.ProcessEnv {
  const clean: NodeJS.ProcessEnv = { ...env };
  delete clean["ANTHROPIC_API_KEY"];
  delete clean["ANTHROPIC_AUTH_TOKEN"];
  return clean;
}

export { SandcastleRunnerError };

File: devclaw/sandcastle_runner.py
"""Per-task docker sandbox runner.

Spawns ``docker run --rm`` against the devclaw-sandbox image for each task. The
container's ENTRYPOINT runs the OpenHands runner (``openhands-runner/runner.py``),
which streams one prefixed JSON line per event (``event: {...}``) plus a single
terminating ``result: {...}`` line. This module:

  - Translates an ``OpenHandsRequest`` into a docker invocation.
  - Bind-mounts the host workspace into /workspace and ~/.claude read-only into
    /home/agent/.claude (Pro OAuth posture: the claude CLI inside the sandbox
    can read tokens but not write back).
  - Streams stdout line-by-line; routes ``event:`` lines through ``on_event``
    and parses the final ``result:`` line as the result.
  - Refuses to forward ANTHROPIC_API_KEY into the container (same belt +
    suspenders the runner enforces).

Container lifecycle: --rm + the per-task --name make destroy-on-exit automatic;
no persistent on-host state. Tests inject a stub runner (via TaskQueue's
``runner`` param) so they don't need docker.
"""

from __future__ import annotations

import asyncio
import json
import os
import uuid
from dataclasses import dataclass
from pathlib import Path
from typing import Awaitable, Callable, Optional, Union

from .state_store import TaskKind

SANDBOX_IMAGE = os.environ.get("DEVCLAW_SANDBOX_IMAGE", "devclaw-sandbox:latest")
DOCKER_BIN = os.environ.get("DEVCLAW_DOCKER_BIN", "docker")
# Container-side mount targets. Match the Dockerfile's expectations.
CONTAINER_WORKSPACE = "/workspace"
CONTAINER_CLAUDE_DIR = "/home/agent/.claude"


@dataclass
class OpenHandsRequest:
    """Inputs the runner needs to launch one OpenHands run. The same kinds the
    MCP tool surface exposes; runner.py picks the right system-prompt per kind."""

    kind: TaskKind
    workspace_dir: str
    goal: str
    #: optional callback, one call per ``event:`` line the runner emits
    on_event: Optional[Callable[["RunnerEvent"], None]] = None


@dataclass
class RunnerEvent:
    id: Optional[str]
    type: str
    source: str
    ts: Union[int, str]
    payload: object


# Terminal verdict from one run. Mirrors the ``result: {...}`` line shape that
# runner.py emits. status == "ok" carries workspace_dir/message (+ agent_output
# for debugging); status == "error" carries error (+ optional trace).
OpenHandsResult = dict


class SandcastleRunnerError(Exception):
    def __init__(self, message: str, trace: str | None = None) -> None:
        super().__init__(message)
        self.trace = trace


def _strip_api_keys(env: dict[str, str]) -> dict[str, str]:
    clean = dict(env)
    clean.pop("ANTHROPIC_API_KEY", None)
    clean.pop("ANTHROPIC_AUTH_TOKEN", None)
    return clean


def _translate_workspace_path(workspace_dir: str) -> str:
    """When devclaw itself runs in a container and spawns docker on the host
    socket, the workspace path it sees internally is not the host's view of
    that bind-mounted dir. The path-prefix env pair tells us how to translate.
    Unset -> pass through (typical local dev, running directly on host)."""
    container_prefix = os.environ.get("DEVCLAW_CONTAINER_PATH_PREFIX")
    host_prefix = os.environ.get("DEVCLAW_HOST_PATH_PREFIX")
    if container_prefix and host_prefix and workspace_dir.startswith(container_prefix):
        return host_prefix + workspace_dir[len(container_prefix) :]
    return workspace_dir


async def run_sandcastle(req: OpenHandsRequest) -> OpenHandsResult:
    """Run one task inside a fresh sandbox container. Resolves with an
    OpenHandsResult dict so it's a drop-in runner for TaskQueue."""
    # DEVCLAW_HOST_CLAUDE_DIR is a HOST path passed straight to docker as a bind
    # source. When devclaw-mcp runs in a container, that path intentionally does
    # NOT exist in the container's view — we pass the string through and let
    # docker emit a clear error if the operator misconfigured the env var.
    claude_dir = os.environ.get("DEVCLAW_HOST_CLAUDE_DIR") or str(
        Path.home() / ".claude"
    )
    host_bind_path = _translate_workspace_path(req.workspace_dir)

    # Per-task container name for greppable logs + manual cleanup if --rm fails.
    container_name = f"devclaw-{uuid.uuid4().hex[:8]}"

    payload = json.dumps(
        {"kind": req.kind, "workspace_dir": CONTAINER_WORKSPACE, "goal": req.goal}
    )

    docker_args = [
        "run",
        "--rm",
        "--name",
        container_name,
        "--network",
        "host",  # claude OAuth refresh needs egress; tighten later via allowlist.
        "-v",
        f"{host_bind_path}:{CONTAINER_WORKSPACE}",
        "-v",
        f"{claude_dir}:{CONTAINER_CLAUDE_DIR}:ro",
        "-e",
        "OPENHANDS_SUPPRESS_BANNER=1",
        SANDBOX_IMAGE,
        payload,
    ]

    try:
        proc = await asyncio.create_subprocess_exec(
            DOCKER_BIN,
            *docker_args,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            env=_strip_api_keys(dict(os.environ)),
        )
    except OSError as exc:
        return {
            "status": "error",
            "error": (
                f"failed to spawn {DOCKER_BIN}: {exc}. "
                "Is docker installed and the socket reachable from this process?"
            ),
        }

    result: Optional[OpenHandsResult] = None
    stderr_chunks: list[bytes] = []

    async def drain_stderr() -> None:
        assert proc.stderr is not None
        async for line in proc.stderr:
            stderr_chunks.append(line)

    stderr_task = asyncio.ensure_future(drain_stderr())

    assert proc.stdout is not None
    async for raw in proc.stdout:
        line = raw.decode("utf-8", "replace").strip()
        if not line:
            continue
        if line.startswith("event: "):
            if req.on_event:
                try:
                    data = json.loads(line[len("event: ") :])
                    req.on_event(
                        RunnerEvent(
                            id=data.get("id"),
                            type=data.get("type", ""),
                            source=data.get("source", ""),
                            ts=data.get("ts", 0),
                            payload=data.get("payload"),
                        )
                    )
                except json.JSONDecodeError as parse_err:
                    # malformed event line — drop, don't crash the run
                    import sys

                    sys.stderr.write(
                        f"sandcastle-runner: dropping malformed event line: {parse_err}\n"
                    )
        elif line.startswith("result: "):
            # first result line wins; ignore anything after
            if result is None:
                try:
                    result = json.loads(line[len("result: ") :])
                except json.JSONDecodeError as parse_err:
                    result = {
                        "status": "error",
                        "error": f"runner emitted unparsable result: {parse_err}",
                        "trace": line,
                    }
        # everything else is decorative sandbox output — drop

    await proc.wait()
    stderr_task.cancel()
    stderr_text = b"".join(stderr_chunks).decode("utf-8", "replace")

    if result is not None:
        return result
    return {
        "status": "error",
        "error": (
            f"sandbox exited {proc.returncode} without a result line. "
            f"stderr tail:\n{stderr_text[-1024:]}"
        ),
    }
